### PR TITLE
feat: on demand COJ pdf

### PR DESCRIPTION
**Pull request by Judge40**
Create an endpoint for generating a Conditions of Joining PDF on demand,
if the PDF already exists the existing copy will be downloaded and returned instead.

The data required for the COJ should be provided via a signed programme membership.

TIS21-6106
TIS21-6123

### CODE REVIEW

File: src/main/java/uk/nhs/hee/tis/trainee/forms/config/FilterConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+   * @return The {@link FilterRegistrationBean} for the registration.
+   */
+  @Bean
+  @Order(1)
+  public FilterRegistrationBean<SignedDataFilter> registerSignedDataFilter(
+      SignedDataFilter filter) {
+    FilterRegistrationBean<SignedDataFilter> registrationBean = new FilterRegistrationBean<>();
+
+    registrationBean.setFilter(filter);
+    registrationBean.addUrlPatterns("/api/coj");
```
Comment (ReubenRobertsHEE) [inline src/main/java/uk/nhs/hee/tis/trainee/forms/config/FilterConfiguration.java:69]: Any particular reason why we only care to trace the CoJ api stuff in xray?

File: src/main/java/uk/nhs/hee/tis/trainee/forms/config/FilterConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.core.annotation.Order;
+import uk.nhs.hee.tis.trainee.forms.config.filter.SignedDataFilter;
+
+/**
+ * A configuration for request filters.
+ */
+@Configuration
+public class FilterConfiguration {
```
Comment (ReubenRobertsHEE) [inline src/main/java/uk/nhs/hee/tis/trainee/forms/config/FilterConfiguration.java:37]: The filter stuff looks to be broadly the same as in the credentials service, so I'm inclined to take the rationale and implementation as correct on that basis - I haven't worked much with filters myself 😅

File: src/test/java/uk/nhs/hee/tis/trainee/forms/api/ConditionsOfJoiningResourceTest.java [hunk 3]
```diff
@@ -0,0 +1,457 @@
...
+            """.formatted(programmeMembershipId, version, Instant.now(), Instant.MIN, Instant.MAX),
+        secretKey);
+
+    String key = "40/forms/coj/" + programmeMembershipId + ".pdf";
+    when(service.getUploadedPdf(key)).thenReturn(Optional.empty());
+
+    byte[] response = "response content".getBytes();
+    Resource resource = mock(Resource.class);
+    when(resource.getContentAsByteArray()).thenReturn(response);
+    when(service.generateConditionsOfJoining(any(), eq(false))).thenReturn(resource);
```
Comment (ReubenRobertsHEE) [inline src/test/java/uk/nhs/hee/tis/trainee/forms/api/ConditionsOfJoiningResourceTest.java:416]: It took me a while to figure out that the `eq(false)` is the important bit of this test 😁

**Review by ReubenRobertsHEE — APPROVED** [review]
Looks solid.

File: src/main/java/uk/nhs/hee/tis/trainee/forms/config/FilterConfiguration.java (hunk 2 above)
Comment (Judge40) [reply src/main/java/uk/nhs/hee/tis/trainee/forms/config/FilterConfiguration.java:37]: Yeah, copy paste with a bit of refactoring to remove credentials specific parts (cached session)

File: src/test/java/uk/nhs/hee/tis/trainee/forms/api/ConditionsOfJoiningResourceTest.java (hunk 3 above)
Comment (Judge40) [reply src/test/java/uk/nhs/hee/tis/trainee/forms/api/ConditionsOfJoiningResourceTest.java:416]: I should probably move that to a verification instead, I'll refactor it when I fix the checkbox issue 👍🏻

**Pull request merged**